### ISSUE TITLE
Apple Silicon support

### ISSUE THREAD

**Issue by ganadist** [robolectric/robolectric#6311]
### Description
When perform robolectric test on Apple silicon, following error is raised

```
    Caused by: com.almworks.sqlite4java.SQLiteException: [-91] cannot load library: java.lang.UnsatisfiedLinkError: /private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: dlopen(/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib, 1): no suitable image found.  Did find:
    	/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: no matching architecture in universal wrapper
    	/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: no matching architecture in universal wrapper
    	at com.almworks.sqlite4java.SQLite.loadLibrary(SQLite.java:97)
    	at com.almworks.sqlite4java.SQLite.getSQLiteVersion(SQLite.java:114)
    	at org.robolectric.shadows.util.SQLiteLibraryLoader.loadFromDirectory(SQLiteLibraryLoader.java:90)
    	at org.robolectric.shadows.util.SQLiteLibraryLoader.doLoad(SQLiteLibraryLoader.java:55)
    	at org.robolectric.shadows.util.SQLiteLibraryLoader.load(SQLiteLibraryLoader.java:39)
    	at org.robolectric.shadows.ShadowSQLiteConnection.nativeOpen(ShadowSQLiteConnection.java:73)
    	at org.robolectric.shadows.ShadowSQLiteConnection.nativeOpen(ShadowSQLiteConnection.java:80)
    	at android.database.sqlite.SQLiteConnection.nativeOpen(SQLiteConnection.java)

....
    Caused by: java.lang.UnsatisfiedLinkError: /private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: dlopen(/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib, 1): no suitable image found.  Did find:
    	/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: no matching architecture in universal wrapper
    	/private/var/folders/xp/t20n_5q54v9d7dz1rt9bk9k40000gp/T/1616114208715-0/libsqlite4java.dylib: no matching architecture in universal wrapper
    	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
    	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2442)
    	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2498)
    	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2694)
    	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2627)
    	at java.base/java.lang.Runtime.load0(Runtime.java:768)
    	at java.base/java.lang.System.load(System.java:1837)
    	at com.almworks.sqlite4java.Internal.tryLoadFromPath(Internal.java:340)
    	at com.almworks.sqlite4java.Internal.loadLibraryX(Internal.java:110)
    	at com.almworks.sqlite4java.SQLite.loadLibrary(SQLite.java:95)
    	... 36 more

```

### Robolectric & Android Version
 JVM : Jetbrains Runtime from Intellij IDEA 2020.3.2 ( OpenJDK 64-Bit Server VM JBR-11.0.9.1.11-1145.77-jcef (build 11.0.9.1+11-b1145.77, mixed mode) )
 Robolectric : 4.5.1
 Android Compile SDK : 30

**Comment by hoisie** [comment]
Yeah it does not look like sqlite4java is super active. I bet a lot of other projects would have this same issue. If someone is able to build the native library for aarch64 I would be open to finding a way to pull it in during the Robolectric sqlite copy native process:

https://github.com/robolectric/robolectric/blob/master/shadows/framework/build.gradle#L20

Seems like getting this working could be a matter of figuring out the sqlite4java native build process.

**Comment by ZacSweers** [comment]
We have an M1 device and happy to help test any fixes!

**Comment by ganadist** [comment]
sqlite4java is using `gant` as build system.
And it is hard to setup on M1 device.

So, I wrote simple [Makefile](https://github.com/ganadist/sqlite4java/blob/arm64_port/Makefile) instead of it.

And here is commit to apply sqlite4java binary for Apple silicon.
* master : https://github.com/ganadist/robolectric/commit/1f75daac70b0a5faff3ab2d5ffed4267a6d1cfc0
* 4.5.x : https://github.com/ganadist/robolectric/commit/cd2e009e61ba6b7e47f36e0148fc343a8b96e14d

These commit has a binary file, so please consider to apply git-lfs on robolectric repository if you want to apply it.

**Comment by ganadist** [comment]
I created a prebuilt binary base robolectric shadows framework on https://github.com/ganadist/robolectric/releases/tag/4.5.1-apple-m1

**Comment by ZacSweers** [comment]
I'll give that a shot and report back

**Comment by hoisie** [comment]
Can you upload it to maven, like the other sqlite natives? That way
Robolectric could pull it in as part of the build process and it would not
have to be checked in.

**Comment by ganadist** [comment]
> Can you upload it to maven, like the other sqlite natives? That way Robolectric could pull it in as part of the build process and it would not have to be checked in.

I created a ticket to publish on maven central.
https://issues.sonatype.org/browse/OSSRH-66040

After published it, I'll reply again.

**Comment by hoisie** [comment]
Yeah the situation with sqlite4java is not great. There are already 2-3
open bugs in it that affect Robolectric.

I think the issue with a generic jdbc sqlite driver is certain Android API
methods seem to require calling specific C APIs on sqlite.

For instance:

https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase#setCustomScalarFunction(java.lang.String,%20java.util.function.UnaryOperator%3Cjava.lang.String%3E)

I am not sure there is a way to implement that using a generic raw query.
To define a scalar function I believe you need to call a specific C method
on the sqlite API.

**Comment by kolsys** [comment]
The package already in Maven 
https://search.maven.org/artifact/io.github.ganadist.sqlite4java/libsqlite4java-osx-arm64/1.0.392/dylib

What about hotfix?

**Comment by hoisie** [comment]
@kolsys I am waiting for someone to verify that this new binary works. I personally do not have an M1.

**Comment by kolsys** [comment]
@hoisie I'm ready to test but don't understand how without org.robolectric:robolectric package.

**Comment by ganadist** [comment]
@kolsys : Here is PR #6324 which can apply https://search.maven.org/artifact/io.github.ganadist.sqlite4java/libsqlite4java-osx-arm64/1.0.392/dylib on robolectric

And you can test with your own built robolectrics binary.
https://github.com/robolectric/robolectric/pull/6324#issuecomment-805019928

In my case, I tested with https://github.com/ganadist/sqldelight

To test master branch of robolectric
```
$ git clone https://github.com/robolectric/robolectric
$ cd robolectric
$ git checkout origin/master
$ git fetch origin refs/pull/6324/head
$ git cherry-pick FETCH_HEAD
$ env SKIP_JAVADOC=true ./gradlew publishMavenJavaPublicationToMavenLocal

$ cd ..
$ git clone https://github.com/ganadist/sqldelight
$ cd sqldelight
$ git checkout origin/local_robolectrics
$ ./gradlew testDebugUnitTest --no-build-cache

BUILD FAILED in 5s
32 actionable tasks: 4 executed, 28 up-to-date

Publishing build scan...
https://gradle.com/s/aec36hfjjurjy

```
Unfortunately, upstream (master branch) version of robolectric currently throws [errors from ShadowLegacyAssetManager](https://scans.gradle.com/s/aec36hfjjurjy/tests/:extensions:android-paging:testDebugUnitTest/com.squareup.sqldelight.android.paging.QueryDataSourceFactoryTest/initial%20load%20gives%20expected%20results%20back?expanded-stacktrace=WyIwLTEtMi0zIl0&top-execution=1)

You can also test with stable version of robolectric

```
$ git clone https://github.com/robolectric/robolectric
$ cd robolectric
$ git checkout robolectric-4.5.1
$ git fetch origin refs/pull/6324/head
$ git cherry-pick FETCH_HEAD
$ env SKIP_JAVADOC=true ./gradlew publishMavenJavaPublicationToMavenLocal

$ cd ..
$ git clone https://github.com/ganadist/sqldelight
$ cd sqldelight
$ git checkout origin/local_robolectrics_4.5
$ ./gradlew testDebugUnitTest --no-build-cache

BUILD SUCCESSFUL in 6s
52 actionable tasks: 4 executed, 48 up-to-date

Publishing build scan...
https://gradle.com/s/j5tak2m2dulwk

```

**Comment by kolsys** [comment]
```
$ ./gradlew testDebugUnitTest --no-build-cache

BUILD FAILED in 36s
45 actionable tasks: 45 executed

Publishing build scan...
https://gradle.com/s/z5tju2oeazl2w
```

In my project I'm also getting errors in some cases but its not look like sqlite3 problems, the okhttp3.mockwebserver.MockWebServer is not works as expected.

**Comment by ganadist** [comment]
> In my project I'm also getting errors in some cases but its not look like sqlite3 problems.
Yes. we cannot verify #6324 against master branch, right now.

**Comment by dockusan** [comment]
👍 
I have the same issue when execute some Android Room test cases on Mac M1.
Switch from arm64 JDK to x86_64 JDK works, but performance issue happen. (take very long time to build).

**Comment by ganadist** [comment]
> I have the same issue when execute some Android Room test cases on Mac M1.

Your issue is not related with robolectric.
Please see https://issuetracker.google.com/issues/174695268#comment10

**Comment by dockusan** [comment]
> > I have the same issue when execute some Android Room test cases on Mac M1.
> 
> Your issue is not related with robolectric.
> Please see https://issuetracker.google.com/issues/174695268#comment10

![Screen Shot 2021-08-02 at 2 01 51 PM](https://user-images.githubusercontent.com/10556072/127817875-8113efc7-a94d-49c5-8f7a-9cb2424f9801.png)
The error point to Robolectric code. 
I'm using RxJava3 to test with Room: 
```
sut.insertLog(log).test().assertComplete()
```

**Comment by vialyx** [comment]
Have the same problem when running tests for Android project.
<img width="1068" alt="Screen Shot 2021-08-20 at 9 17 09 AM" src="https://user-images.githubusercontent.com/8024661/130195316-29d4546d-17d7-4ca6-b165-36a77dc22d79.png">
Ready to test early versions of fix.

**Comment by hoisie** [comment]
Just to update this, there are some new experimental SQLite shadows being developed that will use native code derived from Android R AOSP. I am totally pegged with a couple other things at the moment, but hopefully later this week I will be able to get the Linux ones working in a gradle/maven environment. If that goes well, my plan is to add support for Mac M1 immediately so folks aren't unblocked by that. Windows users will probably have to use the legacy SQLite shadows at the moment. Hopefully once Linux and Mac are supported we can tackle Windows.

**Comment by hoisie** [comment]
OK as of #6720, Robolectric builds native SQLite artifacts for Linux x86_64. I have a M1 Mac Mini and will set it up to try to get it working for that as well.

This is where the sqlite jni code is:
https://github.com/robolectric/robolectric/tree/master/nativeruntime/cpp/jni

Robolectric 4.7 will have a `SQLiteMode(NATIVE)` that lets you use these. It's something like 70% faster than the sqlite4java-backed shadows, and higher fidelity, as you have access to Android collators (UNICODE, LOCALIZED) and extensions like phone_number_equals.

**Comment by mreichelt** [comment]
Is there any way I could help?

I also have an M1 machine and I'm faced with the problem that sqlite4java is not available for arm64. Right now I am using the x86_64 JDK for running the tests, though that really slows down everything.

I fixed some robolectric compile errors and build a `libnativeruntime.dylib` file on my M1 machine. I know it's hacky (I'm not used to C++ development), but maybe this diff could be of a little help? I hereby share it as public domain: 
[nativeruntime_for_mac_arm64.diff](https://github.com/robolectric/robolectric/files/7338652/nativeruntime_for_mac_arm64.diff.txt)

(Note that I had to run cmake & make manually somehow, as the gradle build would always exit with `make: *** No targets specified and no makefile found.  Stop.`)

I had to add ICU to cmake (installed via `brew install icu4c` on my dev machine), and I had to remove the `-static-libgcc` flag, as that seems to be unavailable in clang. I guess that will have to be cleaned up later (maybe by using clang's `-static` flag?).

@hoisie do you think this is helpful - and is there anything else I could help with?

**Comment by mreichelt** [comment]
Nevermind - I just found out you already have [nearly the same changes](https://github.com/hoisie/robolectric/commit/138a513160bfa671fd97260510d79d649d74ea68) on your [m1mini](https://github.com/hoisie/robolectric/commits/m1mini) Branch. So I guess you got this running already 😅

**Comment by bopbi** [comment]
@hoisie based on https://circleci.com/docs/2.0/runner-overview/#supported it seems circleci already have m1 support

**Comment by hoisie** [comment]
@bopbi we have moved to GitHub actions 😢 It is unfortuante that GitHub doesn't even support self-hosted runners yet (https://github.com/actions/runner/issues/805).

**Comment by bopbi** [comment]
hmm, thanks a lot for the info about it

**Comment by hoisie** [comment]
4.7-alpha-1 was just released, which *should* support Apple Silicon. Can folks give it a try and see if there are any issue?

**Comment by yongce** [comment]
4.7-alpha-1 works now.

**Comment by bopbi** [comment]
its working flawlessly on my app also

**Comment by mreichelt** [comment]
And I can confirm as well that `4.7-alpha-1` works on my M1 machine with SQLiteMode native - congrats @hoisie! 👏 

Also, I saw that all architectures (arm64 + x86_64 for Mac, x86_64 for Linux) are all there. Nice! Did anyone check if these now also work on Linux + Intel Macs?

**Comment by ZacSweers** [comment]
When trying the new native support on our linux CI, it seems to require a newer `GLIBC_2.28` version than what comes on AWS boxes normally. Can file a separate issue if need be. Apple silicon does appear to work locally though 👍

**Comment by hoisie** [comment]
Sure, please file a new issue.

**Comment by hoisie** [comment]
@ZacSweers i think I already know what the Linux issue is, I should be building the Linux native libraries with something like Ubuntu 18, I used my workstation which has a more recent version.

**Comment by ZacSweers** [comment]
yup we use ubuntu 18, which makes sense

**Comment by mreichelt** [comment]
@ZacSweers is this the issue that you are mentioning? It seems to fail on our CI Linux machines right now:
```
/usr/local/openjdk-11/bin/java: symbol lookup error: /tmp/17954384001778536976librobolectric-nativeruntime.so: undefined symbol: ucol_open_67
```

**Comment by ZacSweers** [comment]
No that looks different

**Comment by hoisie** [comment]
@mreichelt hmm looks like for some reason libicu is not statically compiled, which is what I intended:

notice the *UND* below:

```
$ objdump -T resources/main/native/linux/x86_64/librobolectric-nativeruntime.so | grep ucol
0000000000000000      D  *UND*	0000000000000000  Base        ucol_strcollIter_67
0000000000000000      D  *UND*	0000000000000000  Base        ucol_close_67
0000000000000000      D  *UND*	0000000000000000  Base        ucol_open_67
0000000000000000      D  *UND*	0000000000000000  Base        ucol_strcoll_67
0000000000000000      D  *UND*	0000000000000000  Base        ucol_setAttribute_67
```

To fix this for now you can install libicu on your ubuntu machine.

I'll try to tweak the CMake file to link against the static version of libicu.

**Comment by hoisie** [comment]
@mreichelt can you give a little more details about your CI environment? Typically most linux machines have libicu installed, it is required for OpenJDK. Do you download a specific JDK without using the system package manager?

**Comment by hoisie** [comment]
Ugh, I spent some time digging into libicu issues, and the situation with libicu Linux is quite complex and convoluted.

The first issue, which is inherent in libicu, is that each library version hardcodes its own version in the symbols, which means that something compiled with libicu v60 on Ubuntu 18 is not compatible with libicu v69 on more recent machines, and vice-versa. This basically means that if you have a .so that has a shared library dependency on libicu, you need the exact version installed, which means that having a shared-library dependency is a non-starter.

So then if you try to depend on the static version of libicu on any Ubuntu version when building a shared object file, you get the error that libicu was not compiled with `-fPIC`, which means you cannot use the system libicu static libraries to build the robolectric nativeruntime shared library.

This means that, when building a release, Robolectric will need to depend on a custom built version of libicu that is compiled with `-fPIC`.

So sadly this will probably require two modes of Linux compilation: if an arbitrary user is building Robolectric, perhaps they can just use the shared-object libicu dependency in order not to have to have a custom libicu dependency. Any version of the Robolectric native runtime built in this mode will likely only work on that machine or machines with an identical profile.

For releases, we will need to have the CI download, configure, and install a custom libICU version, and use a bunch of obscure ld flags to embed the static version. Sadly this will complicate the CMake files considerably :(

**Comment by hoisie** [comment]
@ZacSweers @mreichelt just pushed 4.7-alpha-2, which I hope fixes the Linux issues. Can you give that a spin?

**Comment by mreichelt** [comment]
Nice, both Mac M1 arm64 and Linux x86_64 work for me with 4.7-alpha-2! 🎉 

I can't provide any more information about the CI env as of this moment, but it seems you already solved the issue :)

**Comment by ZacSweers** [comment]
4.7-alpha-2 has everything working for us as well 👍

**Comment by hoisie** [comment]
Marking fixed, 4.7 was just released :)
